spinalcordtoolbox/math.py

```diff
@@ -345,3 +345,99 @@ def denoise_patch2self(data_in, bvals_in, patch_radius=0, model='ols'):
     denoised = dipy.denoise.patch2self.patch2self(data_in, bvals_in, patch_radius=patch_radius, model=model)
 
     return denoised
+
+
+def keep_largest_object(predictions):
+    """Keep the largest connected object from the input array (2D or 3D).
+
+    Taken from:
+    https://github.com/ivadomed/ivadomed/blob/1fccf77239985fc3be99161f9eb18c9470d65206/ivadomed/postprocessing.py#L99-L116
+
+    Args:
+        predictions (ndarray or nibabel object): Input segmentation. Image could be 2D or 3D.
+
+    Returns:
+        ndarray or nibabel (same object as the input).
+    """
+    # Find number of closed objects using skimage "label"
+    labeled_obj, num_obj = label(np.copy(predictions))
+    # If more than one object is found, keep the largest one
+    if num_obj > 1:
+        # Keep the largest object
+        predictions[np.where(labeled_obj != (np.bincount(labeled_obj.flat)[1:].argmax() + 1))] = 0
+    return predictions
+
+
+def fill_holes(predictions, structure=(3, 3, 3)):
+    """Fill holes in the predictions using a given structuring element.
+    Note: This function only works for binary segmentation. 
+
+    Taken from: 
+    https://github.com/ivadomed/ivadomed/blob/master/ivadomed/postprocessing.py#L143
+
+    Args:
+        predictions (ndarray or nibabel object): Input binary segmentation. Image could be 2D or 3D.
+        structure (tuple of integers): Structuring element, number of ints equals
+            number of dimensions in the input array.
+
+    Returns:
+        ndrray or nibabel (same object as the input). Output type is int.
+    """
+    assert np.array_equal(predictions, predictions.astype(bool))
```

Comment:
This is the check that will fail if the input is not binary. However, it doesn't really fail with a useful error message? 

Since people may try to use `-fill-holes` with softseg, we should probably turn this into a useful error message that explains that they need to use `-thr` to binarize their data.

Reply:
good point! added in https://github.com/spinalcordtoolbox/spinalcordtoolbox/pull/4482/commits/3599cb8d40990bffdb0b0a6d35118dc82a18ed2e